File: pages/api/predict.js
// Hapus 'spawn' dan 'path' karena tidak lagi digunakan
// import { spawn } from 'child_process'
// import path from 'path'

// Fungsi fallback Anda (generateMockPrediction) tetap sama, tidak perlu diubah
function generateMockPrediction(inputData) {
  console.log('Generating mock prediction for:', inputData)
  
  let riskScore = 0;
  
  // Age factor
  if (inputData.Age && (inputData.Age < 25 || inputData.Age > 55)) riskScore += 0.2;
  
  // Distance factor
  if (inputData.DistanceFromHome && inputData.DistanceFromHome > 20) riskScore += 0.15;
  
  // Income factor
  if (inputData.MonthlyIncome && inputData.MonthlyIncome < 3000) riskScore += 0.25;
  
  // Overtime factor
  if (inputData.OverTime === 'Yes') riskScore += 0.2;
  
  // Job satisfaction factor
  if (inputData.JobSatisfaction && inputData.JobSatisfaction <= 2) riskScore += 0.3;
  
  // Work-life balance factor
  if (inputData.WorkLifeBalance && inputData.WorkLifeBalance <= 2) riskScore += 0.25;
  
  // Years at company factor (too short or too long)
  if (inputData.YearsAtCompany && (inputData.YearsAtCompany < 1 || inputData.YearsAtCompany > 20)) riskScore += 0.15;
  
  const probability = Math.min(riskScore, 0.95);
  const prediction = probability > 0.5 ? 1 : 0;
  
  const riskLevel = probability > 0.7 ? "High" : probability > 0.4 ? "Medium" : "Low";
  
  return {
    success: true,
    prediction: prediction,
    prediction_label: prediction === 1 ? "Will Leave" : "Will Stay",
    probability: {
      will_stay: 1 - probability,
      will_leave: probability
    },
    risk_level: riskLevel,
    confidence: Math.max(probability, 1 - probability),
    model_type: "Rule-Based Fallback Model",
    feature_importance: {
      "Job Satisfaction": inputData.JobSatisfaction <= 2 ? 0.3 : 0,
      "Work Life Balance": inputData.WorkLifeBalance <= 2 ? 0.25 : 0,
      "Monthly Income": inputData.MonthlyIncome < 3000 ? 0.25 : 0,
      "Over Time": inputData.OverTime === 'Yes' ? 0.2 : 0,
      "Age": (inputData.Age < 25 || inputData.Age > 55) ? 0.2 : 0,
      "Distance From Home": inputData.DistanceFromHome > 20 ? 0.15 : 0
    }
  };
}


// --- INI ADALAH HANDLER API YANG BARU ---
export default async function handler(req, res) {
  console.log('API predict (Node.js) called with method:', req.method);
  console.log('Request headers:', req.headers);
  console.log('Request body type:', typeof req.body);
  console.log('Request body:', req.body);
  
  if (req.method !== 'POST') {
    console.log('Method not allowed - returning 405');
    return res.status(405).json({ 
      success: false, 
      error: 'Method not allowed. Use POST.',
      received_method: req.method,
      debug_info: {
        headers: req.headers,
        body: req.body
      }
    });
  }

  try {
    const inputData = req.body;
    console.log('Input data received:', inputData);

    // Validasi inputData Anda tetap di sini
    const requiredFields = ['Age', 'DistanceFromHome', 'MonthlyIncome', 'YearsAtCompany'];
    for (const field of requiredFields) {
      if (!(field in inputData) || inputData[field] === '' || inputData[field] === null) {
        return res.status(400).json({
          success: false,
          error: `Missing required field: ${field}`
        });
      }
    }

    // --- TRY TO CALL PYTHON API FIRST ---
    
    try {
      // Try to call the Python API
      console.log('Attempting to call Python API...');
      
      const pythonApiUrl = `/api/python/predict`;
      console.log('Calling Python ML API at:', pythonApiUrl);

      const response = await fetch(`${req.headers.origin || 'https://dashboard-business-attrition-vercel-lilac.vercel.app'}${pythonApiUrl}`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(inputData)
      });

      if (response.ok) {
        const result = await response.json();
        console.log('Python API response:', result);

        if (result.success) {
          // Success! Return result from Python ML API
          return res.status(200).json({
            ...result,
            note: 'Prediction from Python ML API'
          });
        } else {
          console.error('Python ML model returned error:', result.error);
          // Fall through to rule-based logic below
        }
      } else {
        const errorText = await response.text();
        console.error('Python API failed:', response.status, errorText);
        // Fall through to rule-based logic below
      }

    } catch (fetchError) {
      console.error('Error calling Python API:', fetchError);
      // Fall through to rule-based logic below
    }

    // --- FALLBACK TO RULE-BASED LOGIC ---
    
    console.log('Using integrated rule-based prediction logic');
    
    // Simple rule-based prediction (same logic as Python API)
    let riskScore = 0.0;
    
    const age = inputData.Age || 30;
    const income = inputData.MonthlyIncome || 5000;
    const overtime = inputData.OverTime || 'No';
    const distance = inputData.DistanceFromHome || 5;
    const jobSatisfaction = inputData.JobSatisfaction || 3;
    const workLifeBalance = inputData.WorkLifeBalance || 3;
    const yearsAtCompany = inputData.YearsAtCompany || 5;
    
    // Apply scoring rules
    if (age < 25 || age > 55) riskScore += 0.3;
    if (income < 3000) riskScore += 0.4;
    if (overtime === 'Yes') riskScore += 0.3;
    if (distance > 20) riskScore += 0.15;
    if (jobSatisfaction <= 2) riskScore += 0.3;
    if (workLifeBalance <= 2) riskScore += 0.25;
    if (yearsAtCompany < 1 || yearsAtCompany > 20) riskScore += 0.15;
    
    const probability = Math.min(riskScore, 0.9);
    const prediction = probability > 0.5 ? 1 : 0;
    
    const result = {
      success: true,
      prediction: prediction,
      prediction_label: prediction === 1 ? "Will Leave" : "Will Stay",
      probability: {
        will_stay: 1 - probability,
        will_leave: probability
      },
      risk_level: probability > 0.7 ? "High" : probability > 0.4 ? "Medium" : "Low",
      confidence: Math.max(probability, 1 - probability),
      model_type: "Integrated Rule-Based Model",
      feature_importance: {
        "Monthly Income": income < 3000 ? 0.4 : 0,
        "Age": (age < 25 || age > 55) ? 0.3 : 0,
        "Over Time": overtime === 'Yes' ? 0.3 : 0,
        "Job Satisfaction": jobSatisfaction <= 2 ? 0.3 : 0,
        "Work Life Balance": workLifeBalance <= 2 ? 0.25 : 0,
        "Distance From Home": distance > 20 ? 0.15 : 0,
        "Years At Company": (yearsAtCompany < 1 || yearsAtCompany > 20) ? 0.15 : 0
      }
    };
    
    console.log('Generated prediction result:', result);
    return res.status(200).json(result);

  } catch (error) {
    // Handle any errors with validation or processing
    console.error('API error in predict.js:', error);
    console.error('With input data:', inputData);
    
    // Fallback ke mock prediction jika terjadi error
    const mockResult = generateMockPrediction(req.body || {});
    return res.status(200).json({
      ...mockResult,
      note: 'Using fallback prediction due to internal server error',
      error_details: error.message
    });
  }
}

File: api/python/predict.py
from http.server import BaseHTTPRequestHandler
import json
import sys
import os

def try_ml_prediction(input_data):
    """Try to use the ML model first"""
    try:
        # Import ML libraries
        from pathlib import Path
        import pandas as pd
        import numpy as np
        import joblib
        
        # Load model and scaler
        model_dir = Path(__file__).parent / 'models'
        model_path = model_dir / 'rf_model.pkl'
        scaler_path = model_dir / 'scaler.pkl'
        
        if not model_path.exists() or not scaler_path.exists():
            return None, "ML model files not found"
        
        model = joblib.load(model_path)
        scaler = joblib.load(scaler_path)
        
        # Process input data for ML model (simplified version)
        # Create a basic feature vector with the most important features
        features = [
            input_data.get('Age', 30),
            input_data.get('MonthlyIncome', 5000), 
            input_data.get('DistanceFromHome', 5),
            input_data.get('YearsAtCompany', 5),
            1 if input_data.get('OverTime', 'No') == 'Yes' else 0,
            input_data.get('JobSatisfaction', 3),
            input_data.get('WorkLifeBalance', 3)
        ]
        
        # For a full 47-feature model, we'd need to pad with defaults
        # For now, let's try with basic features and see if it works
        feature_array = np.array(features).reshape(1, -1)
        
        # This might fail if the model expects 47 features
        # That's okay, we'll fall back to rule-based
        scaled_features = scaler.transform(feature_array)
        prediction = model.predict(scaled_features)[0]
        prediction_proba = model.predict_proba(scaled_features)[0]
        
        return {
            "success": True,
            "prediction": int(prediction),
            "prediction_label": "Will Leave" if prediction == 1 else "Will Stay",
            "probability": {
                "will_stay": float(prediction_proba[0]),
                "will_leave": float(prediction_proba[1])
            },
            "risk_level": "High" if prediction_proba[1] > 0.7 else "Medium" if prediction_proba[1] > 0.4 else "Low",
            "confidence": float(max(prediction_proba)),
            "model_type": "Random Forest ML Model (Python)"
        }, None
        
    except Exception as e:
        return None, f"ML model error: {str(e)}"

def simple_rule_based_prediction(input_data):
    """Fallback rule-based prediction"""
    try:
        age = input_data.get('Age', 30)
        income = input_data.get('MonthlyIncome', 5000)
        overtime = input_data.get('OverTime', 'No')
        job_satisfaction = input_data.get('JobSatisfaction', 3)
        work_life_balance = input_data.get('WorkLifeBalance', 3)
        distance = input_data.get('DistanceFromHome', 5)
        years_at_company = input_data.get('YearsAtCompany', 5)
        
        # Calculate risk score
        risk_score = 0.0
        
        if age < 25 or age > 55:
            risk_score += 0.2
        if income < 3000:
            risk_score += 0.25
        if overtime == 'Yes':
            risk_score += 0.2
        if job_satisfaction <= 2:
            risk_score += 0.3
        if work_life_balance <= 2:
            risk_score += 0.25
        if distance > 20:
            risk_score += 0.15
        if years_at_company < 1 or years_at_company > 20:
            risk_score += 0.15
        
        probability = min(risk_score, 0.95)
        prediction = 1 if probability > 0.5 else 0
        
        return {
            "success": True,
            "prediction": prediction,
            "prediction_label": "Will Leave" if prediction == 1 else "Will Stay",
            "probability": {
                "will_stay": 1 - probability,
                "will_leave": probability
            },
            "risk_level": "High" if probability > 0.7 else "Medium" if probability > 0.4 else "Low",
            "confidence": max(probability, 1 - probability),
            "model_type": "Rule-Based Model (Python Serverless)"
        }
    except Exception as e:
        return {
            "success": False,
            "error": f"Rule-based prediction failed: {str(e)}"
        }

def ml_prediction(input_data):
    """Try ML prediction with pandas/sklearn"""
    try:
        import pandas as pd
        import numpy as np
        import joblib
        from pathlib import Path
        from sklearn.preprocessing import StandardScaler
        
        # Load model
        model_dir = Path(__file__).parent / 'models'
        model_path = model_dir / 'rf_model.pkl'
        scaler_path = model_dir / 'scaler.pkl'
        
        if not model_path.exists() or not scaler_path.exists():
            return None  # Fall back to rule-based
        
        model = joblib.load(model_path)
        scaler = joblib.load(scaler_path)
        
        # Feature engineering (simplified version)
        expected_columns = [
            'EmployeeId', 'Age', 'DailyRate', 'DistanceFromHome', 'Education', 
            'EmployeeCount', 'EnvironmentSatisfaction', 'HourlyRate', 'JobInvolvement',
            'JobLevel', 'JobSatisfaction', 'MonthlyIncome', 'MonthlyRate',
            'NumCompaniesWorked', 'PercentSalaryHike', 'PerformanceRating',
            'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
            'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
            'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
            'YearsWithCurrManager',
            'BusinessTravel_Travel_Frequently', 'BusinessTravel_Travel_Rarely',
            'Department_Research & Development', 'Department_Sales',
            'EducationField_Life Sciences', 'EducationField_Marketing',
            'EducationField_Medical', 'EducationField_Other', 'EducationField_Technical Degree',
            'Gender_Male',
            'JobRole_Human Resources', 'JobRole_Laboratory Technician', 'JobRole_Manager',
            'JobRole_Manufacturing Director', 'JobRole_Research Director', 'JobRole_Research Scientist',
            'JobRole_Sales Executive', 'JobRole_Sales Representative',
            'MaritalStatus_Married', 'MaritalStatus_Single',
            'OverTime_Yes'
        ]
        
        # Create feature dataframe
        result_df = pd.DataFrame(0, index=[0], columns=expected_columns)
        
        # Fill numerical features
        numerical_features = [
            'EmployeeId', 'Age', 'DailyRate', 'DistanceFromHome', 'Education',
            'EmployeeCount', 'EnvironmentSatisfaction', 'HourlyRate', 'JobInvolvement',
            'JobLevel', 'JobSatisfaction', 'MonthlyIncome', 'MonthlyRate',
            'NumCompaniesWorked', 'PercentSalaryHike', 'PerformanceRating',
            'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
            'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
            'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
            'YearsWithCurrManager'
        ]
        
        for col in numerical_features:
            if col in input_data:
                result_df[col] = float(input_data[col]) if input_data[col] not in [None, ''] else 0.0
        
        # Handle categorical features
        if input_data.get('OverTime') == 'Yes':
            result_df['OverTime_Yes'] = 1
        if input_data.get('Gender') == 'Male':
            result_df['Gender_Male'] = 1
            
        # Scale and predict
        scaled_data = scaler.transform(result_df)
        prediction = model.predict(scaled_data)[0]
        prediction_proba = model.predict_proba(scaled_data)[0]
        
        return {
            "success": True,
            "prediction": int(prediction),
            "prediction_label": "Will Leave" if prediction == 1 else "Will Stay",
            "probability": {
                "will_stay": float(prediction_proba[0]),
                "will_leave": float(prediction_proba[1])
            },
            "risk_level": "High" if prediction_proba[1] > 0.7 else "Medium" if prediction_proba[1] > 0.4 else "Low",
            "confidence": float(max(prediction_proba)),
            "model_type": "Random Forest ML Model"
        }
        
    except Exception as e:
        print(f"ML prediction failed: {e}", file=sys.stderr)
        return None  # Fall back to rule-based

class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        try:
            # Read the request body
            content_length = int(self.headers['Content-Length'])
            post_data = self.rfile.read(content_length)
            
            # Parse JSON data
            input_data = json.loads(post_data.decode('utf-8'))
            
            # Validate required fields
            required_fields = ['Age', 'DistanceFromHome', 'MonthlyIncome', 'YearsAtCompany']
            for field in required_fields:
                if field not in input_data or input_data[field] == '' or input_data[field] is None:
                    self.send_response(400)
                    self.send_header('Content-Type', 'application/json')
                    self.send_header('Access-Control-Allow-Origin', '*')
                    self.end_headers()
                    
                    error_result = {
                        "success": False,
                        "error": f"Missing required field: {field}"
                    }
                    self.wfile.write(json.dumps(error_result).encode('utf-8'))
                    return
            
            # Try ML prediction first
            ml_result, ml_error = try_ml_prediction(input_data)
            
            if ml_result and ml_result.get('success'):
                result = ml_result
                result["note"] = "Using Random Forest ML model"
            else:
                # Fall back to rule-based
                result = simple_rule_based_prediction(input_data)
                if ml_error:
                    result["note"] = f"Using rule-based prediction (ML error: {ml_error})"
                else:
                    result["note"] = "Using rule-based prediction (ML model not available)"
            
            # Send response
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
            self.send_header('Access-Control-Allow-Headers', 'Content-Type')
            self.end_headers()
            
            self.wfile.write(json.dumps(result).encode('utf-8'))
            
        except Exception as e:
            # Handle errors
            error_result = {
                "success": False,
                "error": f"Internal server error: {str(e)}",
                "error_type": type(e).__name__
            }
            
            self.send_response(500)
            self.send_header('Content-Type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            
            self.wfile.write(json.dumps(error_result).encode('utf-8'))
    
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()